File: timerx-core/src/main/java/timerx/format/Semantic.java
package timerx.format;

import androidx.annotation.NonNull;

public class Semantic {

  Position hoursPosition;
  Position minutesPosition;
  Position secondsPosition;
  Position rMillisPosition;
  TimeUnitType smallestAvailableUnit;

  String format;
  String strippedFormat;

  Semantic(@NonNull Position hoursPosition, @NonNull Position minutesPosition,
      @NonNull Position secondsPosition, @NonNull Position rMillisPosition,
      String format, String strippedFormat,
      TimeUnitType smallestAvailableUnit) {
    this.hoursPosition = hoursPosition;
    this.minutesPosition = minutesPosition;
    this.secondsPosition = secondsPosition;
    this.rMillisPosition = rMillisPosition;
    this.format = format;
    this.strippedFormat = strippedFormat;
    this.smallestAvailableUnit = smallestAvailableUnit;
  }

  @NonNull
  public String getFormat() {
    return strippedFormat;
  }

  boolean has(@NonNull TimeUnitType unitType) {
    switch (unitType) {
      case HOURS:
        return hoursPosition.isNotEmpty();
      case MINUTES:
        return minutesPosition.isNotEmpty();
      case SECONDS:
        return secondsPosition.isNotEmpty();
      case R_MILLISECONDS:
        return rMillisPosition.isNotEmpty();
    }
    throw new IllegalArgumentException("Incorrect type of unit");
  }

  boolean hasOnlyRMillis() {
    return (rMillisPosition.isNotEmpty() && secondsPosition.isEmpty()
        && minutesPosition.isEmpty() && hoursPosition.isEmpty());
  }
}

File: timerx-core/src/main/java/timerx/Semantic.kt
package timerx

/**
 * Information about format, such as positions of different units, smallest available unit
 * and stripped version of format without escape symbols
 */
internal class Semantic(
  val hoursPosition: Position,
  val minutesPosition: Position,
  val secondsPosition: Position,
  val rMillisPosition: Position,
  val format: String,
  val strippedFormat: String,
  val smallestAvailableUnit: TimeUnitType,
  val largestAvailableUnit: TimeUnitType
) {
  
  val largestAvailableUnitLength: Int
    get() = getPositionOf(largestAvailableUnit).length
  
  fun has(unitType: TimeUnitType) = when (unitType) {
    TimeUnitType.HOURS -> hoursPosition.isNotEmpty
    TimeUnitType.MINUTES -> minutesPosition.isNotEmpty
    TimeUnitType.SECONDS -> secondsPosition.isNotEmpty
    TimeUnitType.R_MILLISECONDS -> rMillisPosition.isNotEmpty
  }
  
  fun hasOnlyRMillis(): Boolean {
    return (rMillisPosition.isNotEmpty && secondsPosition.isEmpty
        && minutesPosition.isEmpty && hoursPosition.isEmpty)
  }
  
  fun getPositionOf(unitType: TimeUnitType) = when (unitType) {
    TimeUnitType.HOURS -> hoursPosition
    TimeUnitType.MINUTES -> minutesPosition
    TimeUnitType.SECONDS -> secondsPosition
    TimeUnitType.R_MILLISECONDS -> rMillisPosition
  }
}